[c10d] Pass avoidRecordStreams into collective() function

kwen2501 opened the pull request:
Even after PR #111431, the `collective(...)` function still uses the underlined version `avoidRecordStreams_` inside and does not respect each collective call's preference, as the underlined `avoidRecordStreams_` is only controlled by environment variable.

As a fix, we pass `avoidRecordStreams` into the collective() function.

kwen2501 committed:
[c10d] Pass avoidRecordStreams into collective() function

awgu reviewed (approved):
This makes sense to me!

awgu commented:
It looks like this broke correctness from the unit tests :/

kwen2501 committed:
Revert changes in _base functions

kwen2501 committed:
Add record_stream missed in FSDP's offload path

The sharded param is created by `pre_unshard` stream, then used by
`unshard` stream, then PG-NCCL's stream. A `record_stream` must be added
between `pre_unshard` and `unshard` if PG-NCCL stops doing such favor.

awgu reviewed (approved):
We discussed the FSDP change offline. FSDP was missing a `record_stream` call for correctness and was relying on `ProcessGroupNCCL`'s own `record_stream` call, which is not technically part of the API's contract.

In other words, this change to `ProcessGroupNCCL` _can_ lead to silent correctness failures, but it would only be for existing code that depended on unspecified behavior.

awgu commented:
@pytorchbot merge
